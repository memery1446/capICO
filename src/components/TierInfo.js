import React, { useState, useEffect } from 'react';
import { ethers } from 'ethers';
import { ICO_ADDRESS } from '../contracts/addresses';
import CapICO from '../contracts/CapICO.json';

const TierInfo = () => {
  const [tiers, setTiers] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    const fetchTiers = async () => {
      if (typeof window.ethereum !== 'undefined') {
        try {
          const provider = new ethers.providers.Web3Provider(window.ethereum);
          const contract = new ethers.Contract(ICO_ADDRESS, CapICO.abi, provider);

          const tierCount = await contract.getTierCount();
          const fetchedTiers = [];

          for (let i = 0; i < tierCount; i++) {
            const tier = await contract.getTier(i);
            fetchedTiers.push({
              minPurchase: ethers.utils.formatEther(tier[0]),
              maxPurchase: ethers.utils.formatEther(tier[1]),
              discount: tier[2].toNumber(),
            });
          }

          setTiers(fetchedTiers);
          setLoading(false);
        } catch (err) {
          console.error('Error fetching tier info:', err);
          setError('Failed to load tier information. Please try again later.');
          setLoading(false);
        }
      }
    };

    fetchTiers();
  }, []);

  if (loading) {
    return <div>Loading tier information...</div>;
  }

  if (error) {
    return <div className="text-red-500">{error}</div>;
  }

  return (
    <div className="bg-white p-6 rounded-lg shadow-md">
      <h2 className="text-2xl font-bold mb-4">Investment Tiers</h2>
      {tiers.length === 0 ? (
        <p>No tiers available.</p>
      ) : (
        <table className="w-full">
          <thead>
            <tr className="bg-gray-100">
              <th className="p-2 text-left">Tier</th>
              <th className="p-2 text-left">Min Purchase (ETH)</th>
              <th className="p-2 text-left">Max Purchase (ETH)</th>
              <th className="p-2 text-left">Discount (%)</th>
            </tr>
          </thead>
          <tbody>
            {tiers.map((tier, index) => (
              <tr key={index} className={index % 2 === 0 ? 'bg-gray-50' : ''}>
                <td className="p-2">{index + 1}</td>
                <td className="p-2">{tier.minPurchase}</td>
                <td className="p-2">{tier.maxPurchase}</td>
                <td className="p-2">{tier.discount}</td>
              </tr>
            ))}
          </tbody>
        </table>
      )}
    </div>
  );
};

export default TierInfo;

