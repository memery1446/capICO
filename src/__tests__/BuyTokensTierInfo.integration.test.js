import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import TierInfo from '../components/TierInfo';

const mockStore = configureStore([thunk]);

jest.mock('ethers', () => ({
  ethers: {
    Contract: jest.fn(() => ({
      cooldownTimeLeft: jest.fn().mockResolvedValue({ toNumber: () => 0 }),
      buyTokens: jest.fn().mockResolvedValue({ wait: jest.fn().mockResolvedValue(true) }),
      balanceOf: jest.fn().mockResolvedValue({ toString: () => '1000000000000000000' }),
    })),
    providers: {
      Web3Provider: jest.fn(() => ({
        getSigner: jest.fn(() => ({
          getAddress: jest.fn().mockResolvedValue('0x1234567890123456789012345678901234567890'),
        })),
      })),
    },
    utils: {
      parseEther: jest.fn(val => val),
    },
  },
}));

describe('TierInfo Integration', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      ico: {
        tokenSymbol: 'TEST',
        tokenPrice: '0.1',
        tokenBalance: '10',
        maxPurchaseAmount: '10',
      },
    });
  });

  it('renders loading state correctly', () => {
    const mockGetTiers = jest.fn(() => new Promise(() => {}));
    render(
      <Provider store={store}>
        <TierInfo getTiers={mockGetTiers} />
      </Provider>
    );

    expect(screen.getByText('Loading tier information...')).toBeInTheDocument();
  });

  it('renders tier information correctly after loading', async () => {
    const mockTiers = [
      { minPurchase: '0', maxPurchase: '1', discount: '0' },
      { minPurchase: '1', maxPurchase: '5', discount: '5' },
      { minPurchase: '5', maxPurchase: '10', discount: '10' },
    ];
    const mockGetTiers = jest.fn().mockResolvedValue(mockTiers);

    render(
      <Provider store={store}>
        <TierInfo getTiers={mockGetTiers} />
      </Provider>
    );

    await waitFor(() => {
      expect(screen.getByText('Investment Tiers')).toBeInTheDocument();
    });
  });

  it('handles malformed tier data correctly', async () => {
    const mockTiers = [
      { minPurchase: 'invalid', maxPurchase: '1', discount: '0' },
      { minPurchase: '1', maxPurchase: null, discount: '5' }
    ];
    const mockGetTiers = jest.fn().mockResolvedValue(mockTiers);
    
    render(
      <Provider store={store}>
        <TierInfo getTiers={mockGetTiers} />
      </Provider>
    );

    await waitFor(() => {
      expect(screen.queryByText('Loading tier information...')).not.toBeInTheDocument();
    });
  });

  it('matches snapshot for loaded state', async () => {
    const mockTiers = [
      { minPurchase: '0', maxPurchase: '1', discount: '0' }
    ];
    const mockGetTiers = jest.fn().mockResolvedValue(mockTiers);
    
    const { container } = render(
      <Provider store={store}>
        <TierInfo getTiers={mockGetTiers} />
      </Provider>
    );
    
    await waitFor(() => {
      expect(screen.queryByText('Loading tier information...')).not.toBeInTheDocument();
    });
    
    expect(container).toMatchSnapshot();
  });
// it('handles API errors gracefully', async () => {
//   const mockGetTiers = jest.fn().mockRejectedValue(new Error('Failed to fetch tiers'));
//   render(
//     <Provider store={store}>
//       <TierInfo getTiers={mockGetTiers} />
//     </Provider>
//   );
  
//   await waitFor(() => {
//     expect(screen.getByText(/Error fetching tiers/i)).toBeInTheDocument();
//   });
//});
  // ... rest of your existing tests ...
});
