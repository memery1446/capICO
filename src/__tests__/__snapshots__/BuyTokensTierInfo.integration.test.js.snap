// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TierInfo Integration matches snapshot for loaded state 1`] = `
<div>
  <div
    class="bg-white p-6 rounded-lg shadow-md"
    data-testid="tier-info"
  >
    <h2
      class="text-2xl font-bold mb-4"
    >
      Investment Tiers
    </h2>
    <p
      class="mb-4"
      data-testid="user-investment"
    >
      Your estimated total investment: 
      1.0000
       ETH
    </p>
    <div
      class="overflow-x-auto"
    >
      <table
        class="min-w-full table-auto"
      >
        <thead>
          <tr
            class="bg-gray-200"
          >
            <th
              class="px-4 py-2"
            >
              Tier
            </th>
            <th
              class="px-4 py-2"
            >
              Min Purchase (ETH)
            </th>
            <th
              class="px-4 py-2"
            >
              Max Purchase (ETH)
            </th>
            <th
              class="px-4 py-2"
            >
              Discount (%)
            </th>
            <th
              class="px-4 py-2"
            >
              Status
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="bg-green-100"
            data-testid="tier-1"
          >
            <td
              class="border px-4 py-2"
            >
              1
            </td>
            <td
              class="border px-4 py-2"
            >
              0
            </td>
            <td
              class="border px-4 py-2"
            >
              1
            </td>
            <td
              class="border px-4 py-2"
            >
              0
            </td>
            <td
              class="border px-4 py-2"
            >
              Current
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <p
      class="mt-4 text-sm text-gray-600"
      data-testid="current-tier"
    >
      Your current tier: 
      1
    </p>
    <p
      class="mt-2 text-sm text-gray-600"
      data-testid="next-tier-requirement"
    >
      Next tier requirement: 
      Max tier reached
    </p>
  </div>
</div>
`;
